third/lprng/lprng/man/lprm.1
.ds VE LPRng-3.6.11
.TH LPRM 1 \*(VE "LPRng"
.ig
lprm.1,v 3.39 1998/03/29 18:37:49 papowell Exp
..
.SH NAME
lprm \- remove jobs from the line printer spooling queue
.SH SYNOPSIS
.B lprm
[
.B \-a
]
[
.BI \-D debugopt
]
[
.BI \-P printer
]
[
.B \-V
]
[
.BI \-U user
]
.ti +5n
[
.IR jobid \|.\|.\|.
]
[
.I all
]
.SH DESCRIPTION
.PP
.I Lprm
will send a request to the LPD server to remove jobs
from a spool queue.
(See
.IR clean (1)
for an
.IR lp (1)
compatible interface.)
.I Lprm
is normally the only method by which a user may remove a job.
The jobs to be removed are specified by the job number or user name;
with no specifiers the first removable in a queue will be
selected for removal.
Usually users may only remove jobs submitted by themselves from
the host which originated the original lpr request.
.PP
The spool queue is searched first for jobs that the user has permissions to
remove,
then for jobs that match any of the specified set of tags.
If no tags are specified,
only the first job in the queue that the user has permissions to remove
will be removed.
If the
.B \-a
flag or the
.I all
spool queue is specified,
.I lprm
will search all available spool queues.
.LP
The
.I all
wildcard jobid is used to match all jobs in a spool queue;
however unless the user has
.I control
permissions  for the spool queue only the user's jobs will be removed.
See the
.BR lpd (1)
man page for details about control permissions.
.PP
A jobid can be a job number,
a user name,
a job identifier,
or a pattern for a
.I glob
based wild card match.
This match is applied to the job identifier information.
.PP
.I Lprm
will announce the names of any files it removes and is silent if
there are no jobs in the queue which match the request list.
If the job being removed is active,
the LPD server will stop printing the job and then
restart printing operations.
.SH EXAMPLES
.TP
Remove the last job I submitted if it is in the queue:
.sp
lprm
.TP
Remove job 25 in spool queue p1:
.sp
lprm -Pt1 25
.TP
Remove job 25 and 30 in spool queue p1:
.sp
lprm -Pt1 25 30
.TP
Remove all of user john's jobs in spool queue p1:
.sp
lprm -Pt1 john
.TP
Remove all jobs in spool queue p1:
.sp
lprm -Pt1 all
.TP
Remove all jobs in all spool queues:
.sp
lprm -a all
.TP
Remove the jobs with identifier nobody@system in spool queue p1.
Note that the quotes around the pattern are needed to suppress
shell glob expansion.
.sp
lprm -Pt1 'nobody@system*'
.SH OPTIONS
.IP "\fB\-a\fR" 5
Remove files from all spool queues available to the user.
.IP "\fB\-P\fIprinter\fR" 5
Specifies printer queue.
If it is not provided,
the default printer is the
value of the PRINTER environment variable,
or the first printer in the
.I printcap
database if it is not defined.
.IP "\fB\-U\fIuser\fR" 5
Explicitly provide a userid for doing the removal to the LPD server.
This can be done only by ROOT or userids listed in the
.I allow_user_setting
configuration or printcap option.
This facility is provided to allow printing front end systems such
as SAMBA to submit job removal requests on behalf of users.
.IP "\fB\-D\fIdebugopts\fR" 5
Debugging is controlled using the
.B \-D
option. This accepts a comma-separated list of debugging
settings. These settings take one of two forms: 
.B facility=value , 
or
.B value 
to set an overall default value.
.IP "\fB\-V\fR" 5
The \-V option prints the version information for the program and
verbose information about activities.
.SH WARNINGS
.PP
Users with
CONTROL (i.e.- administrative) permissions on spool queues
can remove any or all jobs.
The
.I all
wildcard defaults to all user jobs,
not just those submitted by the user.
This allows administrators to purge a spool queue easily.
.SH FILES
.PP
The files used by LPRng are set by values in the
printer configuration file.
The following are a commonly used set of default values.
.nf
.ta \w'/var/spool/lpd/printcap.<hostname>           'u
/etc/lpd.conf		LPRng configuration file
/etc/printcap		printer description file
/etc/lpd.perms	printer permissions
/var/spool/printer*		spool directories
/var/spool/printer*/printer	lock file for queue control
/var/spool/printer*/control.printer	queue control
/var/spool/printer*/active.printer	active job
/var/spool/printer*/log.printer	log file
.fi
.SH "SEE ALSO"
lpd.conf(5),
lpc(8),
lpd(8),
lpr(1),
lpq(1),
lprm(1),
printcap(5),
lpd.conf(5),
pr(1).
.SH DIAGNOSTICS
.nf
Most of the diagnostics are self explanatory.
If you are puzzled over the exact cause of failure,
set the debugging level on (-D5) and run again.
The debugging information will 
help you to pinpoint the exact cause of failure.
.fi
.SH "HISTORY"
LPRng is a enhanced printer spooler system
with functionality similar to the Berkeley LPR software.
The LPRng mailing list is lprng@lprng.com;
subscribe by sending mail to lprng-request@lprng.com with
the word subscribe in the body.
The software is available from ftp://ftp.astart.com/pub/LPRng.
.SH "AUTHOR"
Patrick Powell <papowell@astart.com>.


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/lprng/lprng/src/common/lprm.c
/***************************************************************************
 * LPRng - An Extended Print Spooler System
 *
 * Copyright 1988-1999, Patrick Powell, San Diego, CA
 *     papowell@astart.com
 * See LICENSE for conditions of use.
 *
 ***************************************************************************/

 static char *const _id =
"$Id: lprm.c,v 1.6.2.1 2001-03-07 01:41:59 ghudson Exp $";


/***************************************************************************
 * LPRng - An Extended Print Spooler System
 *
 * Copyright 1988-1997, Patrick Powell, San Diego, CA
 *     papowell@astart.com
 * See LICENSE for conditions of use.
 *
 ***************************************************************************
 * MODULE: lprm.c
 * PURPOSE:
 **************************************************************************/

/***************************************************************************
 * SYNOPSIS
 *      lprm [ -PPrinter_DYN ]
 *    lprm [-Pprinter ]*[-a][-s][-l][+[n]][-Ddebugopt][job#][user]
 * DESCRIPTION
 *   lprm sends a status request to lpd(8)
 *   and reports the status of the
 *   specified jobs or all  jobs  associated  with  a  user.  lprm
 *   invoked  without  any arguments reports on the printer given
 *   by the default printer (see -P option).  For each  job  sub-
 *   mitted  (i.e.  invocation  of lpr(1)) lprm reports the user's
 *   name, current rank in the queue, the names of files compris-
 *   ing  the job, the job identifier (a number which may be sup-
 *   plied to lprm(1) for removing a specific job), and the total
 *   size  in  bytes.  Job ordering is dependent on the algorithm
 *   used to scan the spooling directory and is  FIFO  (First  in
 *   First Out), in order of priority level.  File names compris-
 *   ing a job may be unavailable (when lpr(1) is used as a  sink
 *   in  a  pipeline)  in  which  case  the  file is indicated as
 *   ``(stdin)''.
 *    -P printer
 *         Specifies a particular printer, otherwise  the  default
 *         line printer is used (or the value of the PRINTER vari-
 *         able in the environment).  If PRINTER is  not  defined,
 *         then  the  first  entry in the /etc/printcap(5) file is
 *         reported.  Multiple printers can be displayed by speci-
 *         fying more than one -P option.
 *
 *   -a   All printers listed in the  /etc/printcap(5)  file  are
 *        reported.
 *
 *   -l   An alternate  display  format  is  used,  which  simply
 *        reports the user, jobnumber, and originating host.
 *
 *   [+[n]]
 *        Forces lprm to periodically display  the  spool  queues.
 *        Supplying  a  number immediately after the + sign indi-
 *        cates that lprm should sleep n seconds in between  scans
 *        of the queue.
 *        Note: the screen will be cleared at the start of each
 *        display using the 'curses.h' package.
 ****************************************************************************
 *
 * Implementation Notes
 * Patrick Powell Tue May  2 09:58:29 PDT 1995
 * 
 * The LPD server will be returning the formatted status;
 * The format can be the following:
 * 
 * SHORT:
 * Warning: lp is down: lp is ready and printing
 * Warning: no daemon present
 * Rank   Owner      Job  Files                                 Total Size
 * active root       30   standard input                        5 bytes
 * 2nd    root       31   standard input                        5 bytes
 * 
 * LONG:
 * 
 * Warning: lp is down: lp is ready and printing
 * Warning: no daemon present
 * 
 * root: 1st                                [job 030taco]
 *         standard input                   5 bytes
 * 
 * root: 2nd                                [job 031taco]
 *         standard input                   5 bytes
 * 
 */

#include "lp.h"
#include "child.h"
#include "getopt.h"
#include "getprinter.h"
#include "getqueue.h"
#include "initialize.h"
#include "linksupport.h"
#include "patchlevel.h"
#include "sendauth.h"
#include "sendreq.h"

/**** ENDINCLUDE ****/

#undef EXTERN
#undef DEFINE
#define EXTERN
#define DEFINE(X) X
#include "lprm.h"
/**** ENDINCLUDE ****/

void Do_removal(char **argv);

/***************************************************************************
 * main()
 * - top level of LPRM
 *
 ****************************************************************************/

int main(int argc, char *argv[], char *envp[])
{
	int i;
	struct line_list args;

	Init_line_list(&args);

	/* set signal handlers */
	(void) plp_signal (SIGHUP, cleanup_HUP);
	(void) plp_signal (SIGINT, cleanup_INT);
	(void) plp_signal (SIGQUIT, cleanup_QUIT);
	(void) plp_signal (SIGTERM, cleanup_TERM);

	/*
	 * set up the user state
	 */
#ifndef NODEBUG
	Debug = 0;
#endif


	Initialize(argc, argv, envp);
	Setup_configuration();
	Get_parms(argc, argv);      /* scan input args */

	Add_line_list(&args,Logname_DYN,0,0,0);
	for( i = Optind; argv[i]; ++i ){
		/* Compatibility with BSD lprm */
		if( !strcmp(argv[i], "-") )
			Add_line_list(&args,Logname_DYN,0,0,0);
		else
			Add_line_list(&args,argv[i],0,0,0);
	}
	Check_max(&args,2);
	args.list[args.count] = 0;

	/* now force the printer list */
	if( All_printers || (Printer_DYN && strcasecmp(Printer_DYN,ALL) == 0 ) ){
		All_printers = 1;
		Get_all_printcap_entries();
		if(DEBUGL1)Dump_line_list("lprm - final All_line_list", &All_line_list);
	}
	DEBUG1("lprm: Printer_DYN '%s', All_printers %d, All_line_list.count %d",
		Printer_DYN, All_printers, All_line_list.count );
	/* we do the individual printers */
	if( All_printers ){
		if( All_line_list.count == 0 ){
			fprintf(stderr,"no printers\n");
			cleanup(0);
		}
		for( i = 0; i < All_line_list.count; ++i ){
			Set_DYN(&Printer_DYN,All_line_list.list[i] );
			Do_removal(args.list);
		}
	} else {
		Do_removal(args.list);
	}
	Free_line_list(&args);
	DEBUG1("lprm: done");
	Remove_tempfiles();
	DEBUG1("lprm: tempfiles removed");
	Errorcode = 0;
	DEBUG1("lprm: cleaning up");
	cleanup(0);
	return(0);
}


void Do_removal(char **argv)
{
	int fd, n;
	char msg[LINEBUFFER];

	DEBUG1("Do_removal: start");
	Get_printer();
	Fix_Rm_Rp_info();
	/* fix up authentication */
	if( Auth_JOB ){
		/* Fix_auth will update Auth_DYN later */
		Set_str_value( &PC_entry_line_list, "auth", Auth_JOB );
	}
	if( Check_for_rg_group( Logname_DYN ) ){
		fprintf( stderr,
			"cannot use printer - not in privileged group\n" );
		return;
	}
	fd = Send_request( 'M', REQ_REMOVE,
		argv, Connect_timeout_DYN, Send_query_rw_timeout_DYN, 1 );
	while( (n = read(fd, msg, sizeof(msg)) ) > 0 ){
		if( write(1,msg,n) < 0 ) cleanup(0);
	}
	close(fd);
	DEBUG1("Do_removal: end");
}


/* VARARGS2 */
#ifdef HAVE_STDARGS
void setstatus (struct job *job,char *fmt,...)
#else
void setstatus (va_alist) va_dcl
#endif
{
#ifndef HAVE_STDARGS
    struct job *job;
    char *fmt;
#endif
	char msg[LARGEBUFFER];
    VA_LOCAL_DECL

    VA_START (fmt);
    VA_SHIFT (job, struct job * );
    VA_SHIFT (fmt, char *);

	msg[0] = 0;
	if( Verbose ){
		(void) plp_vsnprintf( msg, sizeof(msg)-2, fmt, ap);
		strcat( msg,"\n" );
		if( Write_fd_str( 2, msg ) < 0 ) cleanup(0);
	}
	VA_END;
	return;
}

void send_to_logger (struct job *job,const char *header, char *fmt){;}

/* VARARGS2 */
#ifdef HAVE_STDARGS
void setmessage (struct job *job,const char *header, char *fmt,...)
#else
void setmessage (va_alist) va_dcl
#endif
{
#ifndef HAVE_STDARGS
    struct job *job;
    char *fmt, *header;
#endif
	char msg[LARGEBUFFER];
    VA_LOCAL_DECL

    VA_START (fmt);
    VA_SHIFT (job, struct job * );
    VA_SHIFT (header, char *);
    VA_SHIFT (fmt, char *);

	msg[0] = 0;
	if( Verbose ){
		(void) plp_vsnprintf( msg, sizeof(msg)-2, fmt, ap);
		strcat( msg,"\n" );
		if( Write_fd_str( 2, msg ) < 0 ) cleanup(0);
	}
	VA_END;
	return;
}


/***************************************************************************
 * void Get_parms(int argc, char *argv[])
 * 1. Scan the argument list and get the flags
 * 2. Check for duplicate information
 ***************************************************************************/

extern char *next_opt;
void usage(void);
char LPRM_optstr[]   /* LPRM options */
 = "A:aD:P:V" ;
char CLEAN_optstr[]   /* CLEAN options */
 = "D:" ;

void Get_parms(int argc, char *argv[] )
{
	int option;
	char *name;


	if( argv[0] && (Name = strrchr( argv[0], '/' )) ) {
		++Name;
	} else {
		Name = argv[0];
	}
	/* check to see if we simulate (poorly) the LP options */
	if( Name && strcmp( Name, "clean" ) == 0 ){
		LP_mode = 1;
		while ((option = Getopt (argc, argv, CLEAN_optstr )) != EOF)
		switch (option) {
		case 'D':
			Parse_debug( Optarg, 1 );
			break;
		default: usage(); break;
		}
		if( Optind < argc ){
			name = argv[argc-1];
			Get_all_printcap_entries();
			if( strcasecmp(name,ALL) ){
				if( Find_exists_value( &All_line_list, name, Value_sep ) ){
					Set_DYN(&Printer_DYN,name);
					argv[argc-1] = 0;
				}
			} else {
				All_printers = 1;
				Set_DYN(&Printer_DYN,"all");
			}
		}
	} else {
		while ((option = Getopt (argc, argv, LPRM_optstr )) != EOF)
		switch (option) {
		case 'A': Auth_JOB = safestrdup(Optarg,__FILE__,__LINE__); break;
		case 'a': All_printers = 1; Set_DYN(&Printer_DYN,"all"); break;
		case 'D': Parse_debug(Optarg, 1); break;
		case 'V': ++Verbose; break;
		case 'P': Set_DYN(&Printer_DYN, Optarg); break;
		default: usage(); break;
		}
	}
	if( Verbose ){
		fprintf( stderr, _("Version %s\n"), PATCHLEVEL );
		if( Verbose > 1 ) Printlist( Copyright, stderr );
	}
}

char *clean_msg = N_("\
usage: %s [-Ddebuglevel] (jobid|user|'all')* [printer]\n\
  -Ddebuglevel - debug level\n\
  user           removes user jobs\n\
  all            removes all jobs\n\
  jobid          removes job number jobid\n\
 Example:\n\
    'clean 30 lp' removes job 30 on printer lp\n\
    'clean'       removes first job on default printer\n\
    'clean all'      removes all your jobs on default printer\n\
    'clean all all'  removes all your jobs on all printers\n\
  Note: lprm removes only jobs for which you have removal permission\n");

char *lprm_msg = N_("\
usage: %s [-A type] [-a | -Pprinter] [-Ddebuglevel] (jobid|user|'all')*\n\
  -a           - all printers\n\
  -A type      - use specified type of authentication\n\
  -Pprinter    - printer (default PRINTER environment variable)\n\
  -Ddebuglevel - debug level\n\
  -V           - show version information\n\
  user           removes user jobs\n\
  all            removes all jobs\n\
  jobid          removes job number jobid\n\
 Example:\n\
    'lprm -Plp 30' removes job 30 on printer lp\n\
    'lprm -a'      removes all your jobs on all printers\n\
    'lprm -a all'  removes all jobs on all printers\n\
  Note: lprm removes only jobs for which you have removal permission\n");

void usage(void)
{
	if( !LP_mode ){
		fprintf( stderr, _(lprm_msg), Name );
	} else {
		fprintf( stderr, _(clean_msg), Name );
	}
	exit(1);
}

int Start_worker( struct line_list *l, int fd )
{
	return(1);
}

#if TEST

#include "permission.h"
#include "lpd.h"
int Send_request(
	int class,					/* 'Q'= LPQ, 'C'= LPC, M = lprm */
	int format,					/* X for option */
	char **options,				/* options to send */
	int connect_timeout,		/* timeout on connection */
	int transfer_timeout,		/* timeout on transfer */
	int output					/* output on this FD */
	)
{
	int i, n;
	int socket = 1;
	char cmd[SMALLBUFFER];

	cmd[0] = format;
	cmd[1] = 0;
	plp_snprintf(cmd+1, sizeof(cmd)-1, RemotePrinter_DYN);
	for( i = 0; options[i]; ++i ){
		n = strlen(cmd);
		plp_snprintf(cmd+n,sizeof(cmd)-n," %s",options[i] );
	}
	Perm_check.remoteuser = "papowell";
	Perm_check.user = "papowell";
	Is_server = 1;
	Job_remove(&socket,cmd);
	return(-1);
}

#endif
